File: simulation.py
# Simulation of Goodwin and Keen models


# libraries
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d

#
# define functions
#

# Phillips curve
def phill(x2,phi0,phi1):
    return phi0/(1-x2)**2+phi1

# Goodwin model
def GW_dxdt(x,t,alpha,beta,delta,nu):
    x1,x2=x
    dxdt=[x1*(phill(x2,phi0,phi1)-alpha),x2*((1-x1)/nu-alpha-beta-delta)]
    return dxdt 

# Investment function
def kappainv(x,kappa0,kappa1,kappa2,kappa4):
    x1,x3=x
    return kappa0+kappa1*np.arctan(kappa2*(1-x1-r*x3)+kappa4)

# Keen model
def Keen_dxdt(x,t,alpha,beta,delta,nu,r):
    x1,x2,x3=x
    k=kappainv([x1,x3],kappa0,kappa1,kappa2,kappa4)
    dxdt=[x1*(phill(x2,phi0,phi1)-alpha),
          x2*(k/nu-alpha-beta-delta),
          x3*(r-k/nu+delta)+k-(1-x1)]
    return dxdt 
    
    
#
# parameters of the models
#
    
    
#Goodwin parameters
phi0=0.0001
phi1=-0.0376
alpha=0.0129
beta=0.0337
delta=0.055
nu=4.7055

#Keen parameters
r=0.03
kappa0=0.5
kappa1=-0.31831
kappa2=-63.989
kappa4=11.9914


#
# parameters of the simulation
#

#initial points
xGW0=[0.6,0.9] #Goodwin
xKeen0=[0.75,0.94,1] #Keen

#time horizon
tfGW= 20
tfKeen= 150
tGW = np.linspace(0, tfGW, 101)
tKeen = np.linspace(0, tfKeen, 101)



#
# plots
#


solGW = odeint(GW_dxdt, xGW0, tGW, args=(alpha,beta,delta,nu))
plt.figure(1)
plt.plot(solGW[:, 0],solGW[:, 1],'k', label=r'$(\omega(t),\lambda(t)$',linewidth=1)
plt.legend(loc='best')
plt.title('Goodwin Model')
plt.xlabel(r'$\omega(t)$')
plt.ylabel(r'$\lambda(t)$')
plt.grid()
plt.show()


solKeen = odeint(Keen_dxdt, xKeen0, tKeen, args=(alpha,beta,delta,nu,r))
plt.figure(2,figsize=(10, 10), dpi=80, facecolor='w', edgecolor='k')
ax = plt.axes(projection ='3d')
ax.xaxis.pane.fill = False
ax.yaxis.pane.fill = False
ax.zaxis.pane.fill = False
ax.plot(solKeen[:, 0],solKeen[:, 1],solKeen[:, 2], 'green',label=r'$(\omega(t),\lambda(t),d(t))$')
ax.legend()
ax.set_title('Keen Model')
ax.set_xlabel(r'$\omega(t)$')
ax.set_ylabel(r'$\lambda(t)$')
ax.set_zlabel(r'$d(t)$')
plt.grid()
plt.show()

File: simulation.jl
#using Pkg
#Pkg.add("DifferentialEquations")
#Pkg.add("Plots")

using DifferentialEquations
using Plots

phi0=0.0001
phi1=-0.0376
phi=[phi0;phi1]
alpha=0.0129
beta=0.0337
delta=0.055
nu=4.7055

p=[phi0;phi1;alpha;beta;delta;nu]

function phill(x,p,t)
    p[1]/(1-x[2])^2+p[2]
end

function GW(dx,x,p,t)
    phill=phill!(x,p,t) # the symbol ! calls function outside
    dx[1] = x[1]*(phill-p[3])
    dx[2] = x[2]*((1-x[1])/p[6]-p[3]-p[4]-p[5])
end


r=0.03
kappa0=0.5
kappa1=-0.31831
kappa2=-63.989
kappa4=11.9914

pkappa=[kappa0;kappa1;kappa2;kappa4;r]
pkeen=[p;pkappa]


function kappainv(x,pkeen,t)
    pkeen[7]+pkeen[8]*atan(pkeen[9]*(1-x[1]-pkeen[11]*x[3])+pkeen[10])
end



# Keen model
function Keen_dxdt(dx,x,pkeen,t)
    k=kappainv(x,pkeen,t)
    phill=phill!(x,pkeen,t)
    dx[1]=x[1]*(phill-pkeen[3])
    dx[2]=x[2]*(k/pkeen[6]-pkeen[3]-pkeen[4]-pkeen[5])
    dx[3]=x[3]*(pkeen[11]-k/pkeen[6]+pkeen[5])+k-(1-x[1])
end


#Goodwin
xGW0 = [0.6;0.9]
tGWspan = (0,20.0)
probGW = ODEProblem(GW,xGW0,tGWspan,p)
solGW = solve(probGW)

plot(solGW,vars=(1,2))


#Keen
xKeen0 = [0.8;0.9;1]
tKeenspan = (0,150.0)
probKeen = ODEProblem(Keen_dxdt,xKeen0,tKeenspan,pkeen)
solKeen = solve(probKeen)

plot(solKeen,vars=(1,2,3))

